File: internal/model/templates.go
package model

import "time"

// TaskTemplate is a reusable blueprint for creating a task together with its
// subtasks. The template's Name doubles as the title of the root task created
// from it. Captured fields mirror the editable task fields a template supports.
type TaskTemplate struct {
	ID          int64
	Name        string
	Description string
	Priority    Priority
	DayPart     DayPart
	Position    int
	Labels      []Label
	Subtasks    []TaskTemplateSubtask
	CreatedAt   time.Time
	UpdatedAt   time.Time
}

// TaskTemplateSubtask is one child task captured in a template.
type TaskTemplateSubtask struct {
	ID          int64
	TemplateID  int64
	Position    int
	Title       string
	Description string
	Priority    Priority
	DayPart     DayPart
	Labels      []Label
	CreatedAt   time.Time
	UpdatedAt   time.Time
}

File: android-native/core/model/src/main/kotlin/ru/tinyops/turboist/core/model/Templates.kt
package ru.tinyops.turboist.core.model

/**
 * A reusable blueprint for a task and its subtasks. [name] doubles as the title
 * of the task created from it.
 *
 * @property position the order templates are offered in.
 */
data class TaskTemplate(
    override val localId: Long = NO_LOCAL_ID,
    override val serverId: Long? = null,
    val name: String,
    val description: String = "",
    val priority: Priority = Priority.NONE,
    val dayPart: DayPart = DayPart.NONE,
    val position: Int = 0,
    val labels: List<Label> = emptyList(),
    val subtasks: List<TaskTemplateSubtask> = emptyList(),
    val createdAt: Long,
    val updatedAt: Long,
) : ReplicaEntity

/**
 * One child task captured in a template. Templates are one level deep: a deeper
 * tree is flattened into this list when a template is cut from an existing task.
 *
 * @property createdAt the server stores timestamps on the subtask row but leaves
 *   them out of the template payload, so a subtask rebuilt from the wire carries
 *   its template's lifetime rather than its own.
 */
data class TaskTemplateSubtask(
    override val localId: Long = NO_LOCAL_ID,
    override val serverId: Long? = null,
    val templateLocalId: Long = NO_LOCAL_ID,
    val position: Int = 0,
    val title: String,
    val description: String = "",
    val priority: Priority = Priority.NONE,
    val dayPart: DayPart = DayPart.NONE,
    val labels: List<Label> = emptyList(),
    val createdAt: Long = 0L,
    val updatedAt: Long = 0L,
) : ReplicaEntity
